38483 step filter and deferred queries.

sgsteph opened the pull request:
These changes implement two types of optimizations when retrieving a huge number of Tasks from Shotgun:
- The ability to narrow down the list of records retrieved from Shotgun with some Step filtering: only Tasks matching the filter are retrieved form Shotgun. The `ShotgunEntityModel` is cleared and refreshed when Step filters change.
- The ability to use two steps queries: a primary request is used to populate top nodes, a secondary is used only when the data is really needed, typically when the user expand a leaf in the primary tree. secondary ShotgunEntityModels are used to run the deferred queries, allowing the data to be easily cached and retrieved.

The default configuration has been updated to use deferred queries, which is our preferred setup moving forward.

A `ShotgunExtendedEntityModel` class was introduced, with some helpers to re-expand a tree after the model has been reset,  and allowing filters to be updated.
A `ShotgunDeferredEntityModel` class was introduced to deal with deferred queries.

As I went through some parts of the code, I removed dead code, added some missing docstring and fixed some which were wrong. I moved as well some of the inline styling into a standard style.qss file.

sgsteph commented:
Totally optional UI polish linked to this round of work:
https://github.com/shotgunsoftware/tk-framework-shotgunutils/pull/78

jfboismenu commented on python/tk_multi_workfiles/file_finder.py:
```diff
@@ -434,78 +456,76 @@ def _filter_publishes(self, sg_publishes, publish_template, valid_file_extension
         try:
             work_fields = context.as_template_fields(work_template, validate=True)
-        except TankError:
+            missing_keys = work_template.missing_keys(work_fields)
```
Dead code ?

sgsteph replied:
yes, thanks!

jfboismenu commented on python/tk_multi_workfiles/file_finder.py:
```diff
@@ -956,7 +982,6 @@ def _task_process_publish_items(self, sg_publishes, environment, name_map, **kwa
     def _task_find_work_files(self, environment, **kwargs):
         """
         """
-        #time.sleep(5)
```
You have found my dirty secret. This is probably from back when I introduced deferred loading for tabs, i.e, not being loaded from disk until in foreground. ;)

jfboismenu commented on python/tk_multi_workfiles/util.py:
```diff
@@ -89,6 +89,20 @@ def value_to_str(value):
+    """
+    # Deal with some known special cases and assume "code" for anything else.
+    return {
+        "Project": "name",
```
Why are we not using the code you introduced in core recently? Is it simply because we want to avoid requiring a really recent version of tk-core? This would be nice to point out, with the date the feature was introduced in core, so someone who sees the code at a later date can decide if the time has come to use the new core version.

sgsteph replied:
yep, adding a comment.

manneohrstrom commented:
Starting CR now - @sgsteph don't forget to do PRs for defcon2/default config as well!

manneohrstrom commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        Returns a unique id for the given Entity retrieved in a deferred query.
+        """
+        # ShotgunModel uses the entity id for leaves, we use the Entity type and
+        # its id to avoid clashes in the various internal caches.
```
are we sharing a space with internal caches? Is that bad encapsulation?

sgsteph replied:
We don't really share any internal cache, but when adding items in the primary model uids are added in `ShotgunQueryModel.___items_by_uid` for fast lookup. Should I update the comment?

manneohrstrom commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        return "%s_%d" % (sg_entity["type"], sg_entity["id"])
+
+    def ensure_data_is_loaded(self, index=None):
+        """
```
This shadows http://developer.shotgunsoftware.com/tk-framework-shotgunutils/shotgun_model.html#shotgun_model.ShotgunModel.ensure_data_is_loaded - we are not calling the base class - I think it's worth pointing this out in the docstring.

In the top level class docstring, i would love if we could include a list of the methods that are you are supposed to use when calling this class and a short code example ! Are there any existing public sg model methods that no longer make sense to use? Should be avoided even?

sgsteph replied:
I don't think there is any existing public sg model method which shouldn't be used. Added some Python code in the top level class docstring.

manneohrstrom commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+
+    def ensure_data_is_loaded(self, index=None):
+        """
+        Ensure all data is loaded in the model, except for deferred queries.
```
please document index parameter.

manneohrstrom commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        to retrieve the given Entity.
+
+        .. note::
+            The same entity can appear multiple times in the hierarchy, the first
```
for example pipeline steps? agree an example would be helpful!

sgsteph replied:
added a comment.

manneohrstrom commented on python/tk_multi_workfiles/entity_models/extended_model.py:
```diff
@@ -0,0 +1,280 @@
+
+class ShotgunExtendedEntityModel(ShotgunEntityModel):
+    """
+    A Shotgun Entity model with updatable filters and a couple of added methods.
```
same here - would be great with more info. Thinking we move this into sg utils at some point, it's great if we do it properly now. Instead of 'a couple of added methods', it would be great with a list of what was added. When is this class useful/preferred over the std models we already have? And a quick code example would be fantastic.

manneohrstrom commented on python/tk_multi_workfiles/entity_models/extended_model.py:
```diff
@@ -0,0 +1,280 @@
+        Called every time an item was added in the model.
+        """
+        super(ShotgunExtendedEntityModel, self)._finalize_item(item)
+        # We need to keep track of which entities are in the model, so we can bail
```
this comment might be better to move/(or copy) to the ctor where we first introduce the variable

manneohrstrom commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+
+    def item_from_entity(self, entity_type, entity_id):
+        """
+        Retrieve the item representing the given entity in the model.
```
what model is this? only the 'main' one? All loaded child models? How does the deferred-ness play into what is being returned?

sgsteph replied:
From the outside world there is a single model. Adding a comment about the deferredness.

manneohrstrom commented on python/tk_multi_workfiles/entity_models/extended_model.py:
```diff
@@ -0,0 +1,280 @@
+        given value list is not present in the model anymore, the last item
+        matched from the value list is returned.
+
+        :param field_value_list: A list of field values for the path from the
```
please add an example of what this may look like!

manneohrstrom commented on python/tk_multi_workfiles/entity_models/extended_model.py:
```diff
@@ -0,0 +1,280 @@
+        # Reverse the list we return.
+        return values[::-1]
+
+    def _get_key_for_field_data(self, field, sg_data):
```
i think this should be moved down to the deferred model - it doesn't seem to be used anywhere in the extended model.

Generally speaking, it would be great with a more more info around this cache - perhaps in the deferred model class docstring? Explaining what is being cached, how it's being different from the caching already taking place in he sg model and why it's needed!

sgsteph replied:
moved.

sgsteph commented on python/tk_multi_workfiles/browser_form.py:
```diff
@@ -148,25 +168,42 @@ def show_user_filtering_widget(self, is_visible):
-            entity_form = EntityTreeForm(model, caption, allow_task_creation, [], parent=self)
+        for caption, step_filter_on, model in entity_models:
+            step_entity_filter = None
+            if model.represents_tasks:
```
This would not work: in the case of deferred queries the primary `model.get_entity_type()` would return "Asset" (or "Shot"), only the deferred query model would return "Task". Hence, this  extra level of abstraction. Arguably a better name than `represents_tasks` could be found.

sgsteph commented on python/tk_multi_workfiles/browser_form.py:
```diff
@@ -148,25 +168,42 @@ def show_user_filtering_widget(self, is_visible):
+            if model.represents_tasks:
+                # Step filtering on the Entity type the Tasks are linked to or
+                # on Tasks.
+                step_entity_filter = step_filter_on or model.get_entity_type()
```
will clarify.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+    @property
+    def deferred_query(self):
+        """
+        :returns: The deferred query for this model, if any.
```
fixed

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        return self._deferred_query
+
+    @property
+    def represents_tasks(self):
```
See my previous comments. This is not as simple as that: for deferred queries, the primary model entity type is not Task, but Shot or Asset or etc...

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        """
+        :returns: True if Step filtering can be used with this model
+        """
+        # If don't have steps in the fields we query from Shotgun, we assume
```
Steps are non project entities, so it's not possible to have this setup, even in "classic" mode.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        """
+        # If don't have steps in the fields we query from Shotgun, we assume
+        # step filtering should be disabled.
+        return "step" in self._deferred_query["hierarchy"]
```
Given that most of this code is designed for Tasks, I think it will be overkill to support other cases here. Or it would mean a lot of changes to support a general case.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        """
+        Trigger an asynchronous refresh of the model
+        """
+        super(ShotgunDeferredEntityModel, self).async_refresh()
```
I clarified it with a comment: we're refreshing the primary cache, not the cache for deferred queries.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+
+        :param context: A Toolkit context.
+        """
+        if not context:
```
We need to talk about moving this (or not) in sgutils. For the `ShotgunExtendedEntityModel`, there is some extra logic to use `context.task` or `context.entity`, it seems better to let the model deal with this logic, rather than doing it outside of it. And using a `context` here helps to understand the scope of this method.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+
+        :returns: A list of Shotgun results, as returned by a Shotgun `find` call.
+        """
+        deferred_query = self._deferred_query
```
added some

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        message.
+        """
+        parent_item = self.item_from_entity(sg_entity["type"], sg_entity["id"])
+        if not parent_item:
```
added a comment

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        if not changed:
+            return
+        parent_item = self.item_from_entity(sg_entity["type"], sg_entity["id"])
+        if not parent_item:
```
added.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        # there anymore.
+        for uid in existing_uids:
+            if uid in refreshed_uids:
+                # Here we could update items from the refreshed data
```
Adding a TODO, I'm not sure what the side effects could be.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        else:
+            item_list = [self.itemFromIndex(index)]
+
+        while item_list:
```
Because we're modifying the list by adding children. Adding a comment.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+                continue
+            if self.canFetchMore(item.index()):
+                self.fetchMore(item.index())
+            for row_i in range(item.rowCount()):
```
Not sure I understand: we do a `fetchMore` just before iterating on children, so they are fetched?

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        super(ShotgunDeferredEntityModel, self).clear()
+        # Get a fresh and empty deferred cache.
+        self._deferred_cache = ShotgunDataHandlerCache()
+        # TODO: clear deferred models?
```
There was another `clear` method a bit higher in the code. :( Removed this one!

sgsteph commented on python/tk_multi_workfiles/entity_models/extended_model.py:
```diff
@@ -0,0 +1,280 @@
+        to retrieve the given Entity.
+
+        .. note::
+            The same entity can appear multiple times in the hierarchy, the first
```
added a comment.

sgsteph commented on python/tk_multi_workfiles/entity_models/extended_model.py:
```diff
@@ -0,0 +1,280 @@
+        if not self.rowCount():
+            return None
+        parent_list = [self.invisibleRootItem()]
+        while parent_list:
```
the list is modified while being iterated over, added a comment.

jfboismenu replied:
oups... seems I didn't play close enough attention.... my bad!

sgsteph commented on python/tk_multi_workfiles/entity_models/extended_model.py:
```diff
@@ -0,0 +1,280 @@
+        """
+        if not self.rowCount():
+            return None
+        parent = self.invisibleRootItem()
```
Not sure I understand your comment: we have a `fetchMore` so we dynamically load in the model items which were cached?

jfboismenu replied:
I'm thinking of something like
Episode->Sequences->Shot->Step->Task. Let's say Shot's children haven't been fetched from Shotgun yet. canFetchMore will return False, but you won't be able to retrieve Step or Task in the next iteration of the loop because it is asynchronously being loaded.

sgsteph replied:
This is the extended model, not the deferred one, so all data is already pulled from SG, but the model is may be not yet populated with them. I think I added a comment on the deferred model explaining that new data is not fetched from SG if not already pulled.

sgsteph commented on info.yml:
```diff
@@ -97,23 +97,40 @@ configuration:
-          - [entity, type_is, Asset]
-          hierarchy: [entity.Asset.sg_asset_type, entity, step, content]
+          hierarchy: [sg_asset_type, code]
+          sub_hierarchy:
```
Old configs will work, I actually have  a mix of old  and new settings in my config. It's not actually a new exclusive mode (one of or the other), but an extension: if you don't have a `sub_hierarchy `, the  model will stop at first (and unique) hierarchy. If you do have one, then a second one is triggered. I'm not sure I understand your worry: if you don't have `sub_hierarchy` everything will behave like before?
Here are my test settings, please note that there is a new `step_filter_on` setting for classic mode as well:
```
settings.tk-multi-workfiles2.launch_at_startup:
  auto_expand_tree: false
  allow_task_creation: true
  create_new_task_hook: "{self}/create_new_task.py"
  custom_actions_hook: "{self}/custom_actions.py"
  entities:
  - caption: Projects
    entity_type: Project
    filters: []
    hierarchy: [name]
    sub_hierarchy:
        entity_type: Shot
        filters:
        link_field: project
        hierarchy: [sg_sequence]
  - caption: Assets (deferred no step)
    entity_type: Asset
    hierarchy: [sg_asset_type, code]
    sub_hierarchy:
        entity_type: Task
        filters:
        link_field: entity
        hierarchy: null
  - caption: Assets (deferred)
    entity_type: Asset
    hierarchy: [sg_asset_type, code]
    sub_hierarchy:
        entity_type: Task
        filters:
        link_field: entity
        hierarchy: [step]
  - caption: Assets Tasks
    entity_type: Task
    step_filter_on: Asset
    filters:
    - [entity, type_is, Asset]
    hierarchy: [entity.Asset.sg_asset_type, entity, step, content]
  - caption: Shots (deferred)
    entity_type: Shot
    filters:
    - [code, is_not, null]
    hierarchy: [sg_sequence, code]
    sub_hierarchy:
        entity_type: Task
        filters:
        link_field: entity
        hierarchy: [step]
  - caption: Shots Tasks
    entity_type: Task
    filters:
    - [entity, type_is, Shot]
    hierarchy: [entity.Shot.sg_sequence, entity, step, content]
```

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+
+shotgun_model = sgtk.platform.import_framework("tk-framework-shotgunutils", "shotgun_model")
+ShotgunEntityModel = shotgun_model.ShotgunEntityModel
+ShotgunDataHandlerCache = shotgun_model.data_handler_cache.ShotgunDataHandlerCache
```
yes, and there is the discussion about moving this code in shotgunutils.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+    """
+    def __init__(self, entity_type, filters, hierarchy, fields, deferred_query, *args, **kwargs):
+        """
+        A Shotgun Entity model which supports two steps for loading data:
```
I moved it up.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        return self._deferred_query["entity_type"] == "Task"
+
+    @property
+    def supports_step_filtering(self):
```
Sorry to be picky myself, but I don't necessarily agree on this one: I think it's better if the name reflects the purpose rather than what it is based on (the implementation). May be there will be cases where something else will have to be checked to allow or disallow step filtering? Happy to change it to any better name if it's not based on the implementation.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        for uid in uids:
+            item = self._get_item_by_unique_id(uid)
+            if item and item.parent():
+                # Let the parent re-fetch children
```
yes, making it clearer!

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        """
+        Load the data for this model and post a refresh.
+
+        :param extra_filters: A list of additional Shotgun filters which are added
```
Will clarify. I could rename the param to `step_filters`? Something which can be used for both the `ShotgunExtendedEntityModel` (where they are applied to the main query) and the `ShotgunDeferredEntityModel` as it does not seem right to have different names for the params depending on the class.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        """
+        Update the filters used by this model.
+
+        A full refresh is triggered by the update if not using deferred queries.
```
This is some leftover from a previous iteration where there was a single class, fixing that!

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        self._deferred_cache = ShotgunDataHandlerCache()
+        for deferred_model in self._deferred_models.itervalues():
+            deferred_model.clear()
+        self._deferred_models = {}
```
Actually, could we just do `self._deferred_models = {}`? I thought it was needed to clear every single model. I can remove it if not.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        loaded from a deferred query.
+
+        :param parent_item: A :class:`ShotgunStandardItem` instance.
+        :param str uid: A unique id for the new item.
```
fixed and removed `uid`.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        refreshed_uids = []
+        current_item = parent_item
+        current_uid = parent_uid
+        for name in hierarchy:
```
please let me know if it's clearer with the docstring.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        return refreshed_uids
+
+    @classmethod
+    def _dummy_place_holder_item_uid(cls, parent_item):
```
fixed, thanks!

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+                # By having a "text" key and using it for the field name, the tree
+                # view will display its contents.
+                sg_data={
+                    "text": message,
```
yes, it is logged. Adding a comment mentioning it.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        sub_entity_type = deferred_model.get_entity_type()
+        name_field = get_sg_entity_name_field(sub_entity_type)
+
+        # Loop over all entities (leaves) in the deferred model.
```
Added some comments, it wasn't even clear to me when I came back to this code :(

sgsteph replied:
actually I revisited the code a bit, should be clearer now.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+        # but this is a private member.
+        deferred_query = self._deferred_query
+        for sub_entity in sub_entities:
+            uids = self._add_deferred_item_hierarchy(
```
One reason for this is that it allows us to retrieve easily all existing deferred uids by having all of them in a dedicated cache. Another reason comes from the fact that uids for leaves are just entity ids, without the entity type. So we could have clashes between entities with the same id but different types. To summarize: it seemed safer and simpler to do it like that.

sgsteph commented on python/tk_multi_workfiles/entity_models/deferred_model.py:
```diff
@@ -0,0 +1,588 @@
+            data_item = self._deferred_cache.take_item(uid)
+            item = self._get_item_by_unique_id(uid)
+            if item:
+                # The item might be already gone because one its parent was
```
one its parents means its direct parent or a parent higher in the tree. Replaced _parent_ with _ancestor_.

sgsteph commented on python/tk_multi_workfiles/entity_tree/entity_tree_form.py:
```diff
@@ -51,22 +50,32 @@ def __init__(self, label, entity):
         :param search_label:        The hint label to be displayed on the search control
         :param allow_task_creation: Indicates if the form is allowed by the app settings to show the
-                                    create task button.
+                                    create Task button.
```
removed the param which is not needed as the models know this.

sgsteph commented:
Alright, the code is ready for another review pass. I think I addressed or answered all comments. I realized that `extra_filters` (list of filters) was used for something which actually is a single filter, not a list of filters. So I changed some param/member names. Let me know if things look better. If you feel strongly about anything I didn't address, or if I missed anything please shout! ;)
Thanks for your feedback!
